libs/sql-parser/src/main/java/io/crate/sql/parser/AstBuilder.java

```diff
@@ -1991,10 +1991,16 @@ public Node visitBitwiseBinary(SqlBaseParser.BitwiseBinaryContext context) {
 
     @Override
     public Node visitConcatenation(SqlBaseParser.ConcatenationContext context) {
+        var left = (Expression) visit(context.left);
+        var right = (Expression) visit(context.right);
+        if (left instanceof NullLiteral && right instanceof NullLiteral) {
+            return NullLiteral.INSTANCE;
+        }
+
         return new FunctionCall(
-            QualifiedName.of("concat"), List.of(
-            (Expression) visit(context.left),
-            (Expression) visit(context.right)));
+            QualifiedName.of("concat"),
+            List.of(left, right)
+        );
```

Comment:
This won't work if accessing columns. E.g. in PostgreSQL:

```
select x || y from unnest(array[null, 'a', null, 'c'], array[null, null, 'b', 'd']) as t (x, y);
 ?column?
----------
 💩
 💩
 💩
 cd
(4 rows)
```


With this change in CrateDB:

```
cr> select x || y from unnest(array[null, 'a', null, 'c'], array[null, null, 'b', 'd']) as t (x, y);
+--------------+
| concat(x, y) |
+--------------+
|              |
| a            |
| b            |
| cd           |
+--------------+
SELECT 4 rows in set (0.009 sec)
```

I'd revert this part, as it is not worth optimizing the uncommon case (`null || null`) and it actually slows down the more common case (working with columns).

We probably need to introduce a `_||` or `_concat` scalar that's similar to the `ConcatFunction` except with changed null handling.

Reply:
Right 🤦 , good catch, thx.

Reply:
I did like you suggested, see https://github.com/crate/crate/pull/16295/commits/babf8e8a7990ec17a25ea742cfdf0f60fcc74bb1